### PR TITLE
OXT-1407: QEMU v4v stream

### PR DESCRIPTION
**Pull request by jandryuk**
Add a chardev v4v stream mode.  This bypasses the handshaking of the
datagram mode.

By default, v4v chardev runs as datagram, connecting back to QMP in dom0: `-chardev v4v,id=id[,port=5100][,domid=0][,localport=15100][,stream=on|off]`

If localport is not specified, it defaults to port + 10000.

### CODE REVIEW

**Comment by jean-edouard** [comment]
Built here: http://openxt-builder.ainfosec.com:8010/builders/openxt/builds/1655

**Comment by jean-edouard** [comment]
LGTM, merging soon.

File: recipes-openxt/qemu-dm/qemu-dm/qmp-v4v-char-driver.patch [hunk 1]
```diff
@@ -271,6 +284,17 @@
...
 +    }
 +
++    if (s->stream) {
++        if (v4v_connect(fd, &s->remote_addr)) {
++            error_setg(errp,
++                       "%s failed to connect v4v socket - err: %d",
++                       V4V_CHARDRV_NAME, errno);
++            close(fd);
++            return;
++	}
```
[the remark this answers is not in the document]
Comment (jandryuk) [reply recipes-openxt/qemu-dm/qemu-dm/qmp-v4v-char-driver.patch]: Thanks for catching this.

File: recipes-openxt/qemu-dm/qemu-dm/qmp-v4v-char-driver.patch [hunk 2]
```diff
@@ -201,26 +208,31 @@
 +                               Error **errp)
 +{
 +    uint64_t domid = qemu_opt_get_number(opts, "domid", 0);
-+    uint64_t port = qemu_opt_get_number(opts, "port", V4V_QH_PORT);
-+    uint64_t localport = qemu_opt_get_number(opts, "localport", 0);
++    /* Sigh, port is defined as a string */
```
[the remark this answers is not in the document]
Comment (jandryuk) [reply recipes-openxt/qemu-dm/qemu-dm/qmp-v4v-char-driver.patch]: All chardevs use the same option types, which surprised me.  port is defined as a string in chardev/char.c, so we have to match.  I assume that is so port=ftp or similar can work.

**Comment by jandryuk** [comment]
I updated https://openxt.atlassian.net/browse/OXT-1407 with validation steps to test v4v stream mode as a serial chardev.

**Comment by eric-ch** [comment]
Ah I thought this would be used for the QMP connection itself, but `-qmp v4v` is hardcoded, so it will always default to open a v4v DGRAM connection with default options. Using the STREAM protocol should render most of the `V4V_MAGIC_{DIS,}CONNECT`, but I digress, this is futur PR's concerns.

Following your test procedure, `v4v_connect()` to raises `EUCLEAN`:
```
qemu-system-i386: Opened V4V chardev
qemu-system-i386: qemu-system-i386:
qemu-system-i386:  -chardev
qemu-system-i386:  v4v,id=ser-v4v,stream,port=5105
qemu-system-i386: : 
qemu-system-i386: [v4v-chardrv] failed to connect v4v socket - err: 107
```
... This is odd, the v4v driver does not return that error afaik, and the `s->remote` structure is initialized correctly anyway.

**Comment by jandryuk** [comment]
Are you sure you have netcat running in Dom0 with `INET_IS_V4V=1 LD_PRELOAD=/usr/lib/libv4v-1.0.so.0 nc -v -l -p 5105`?  We need to have the v4v port open in Dom0.  I hit a 107 error ENOTCONN myself coming up with the test procedures.  I think it was from the lack of INET_IS_V4V=1.

**Comment by jandryuk** [comment]
(Yes, QMP over stream would be nice, but I haven't gotten there yet.)

**Comment by jandryuk** [comment]
Re-pushed with the whitespace fix.

**Comment by eric-ch** [comment]
Will merge soon.

**Pull request merged**